### DISCUSSION TITLE
Consider some sort of way to have psuedo-explicit abstract class implementations

### DISCUSSION THREAD

**Discussion by john-h-k** [dotnet/csharplang#3445] [Language Ideas]
The idea is just being able to use an explicit interface style implementation on abstract classes. E.g
```cs
public class FrameAllocator<T> : MemoryPool<T>
{
    public struct Frame : IMemoryOwner<T>
    {
        // ...
    }

    public IMemoryOwner<T> MemoryPool<T>.Rent(int size) => Rent(size);
    public Frame Rent(int size) => new Frame(size);
}
```

This would allow boxing to be avoided when calling exact types

**Comment by HaloFour** [comment]
This looks like a scenario that would be satisfied by covariant return types: 
 https://github.com/dotnet/csharplang/issues/49

**Comment by john-h-k** [comment]
It isn't
> The override method must have have a return type that is convertible by an identity or implicit reference conversion to the return type of the overridden base method.

Boxing is not an identity conversion

**Comment by svick** [comment]
A very similar proposal: https://github.com/dotnet/csharplang/issues/1618.

**Comment by HaloFour** [comment]
Good point, I missed that `Frame` is a `struct` in that example.

The CLR does support the concept of explicit overrides.  As long as the signatures match the overriding method can have any name or accessibility.

**Comment by ViIvanov** [comment]
Nice thing, but not obvious, how to derived class can override `MemoryPool<T>::Rent(int)` and `FrameAllocator<T>::Rent(int)` (in case this is virtual method).

// This is why I prefer "non-virtual interface" design.

**Comment by Pentadome** [comment]
Couldn't you just make a "ValueRent"` method?

**Comment by john-h-k** [comment]
Yes, and that is my current solution. But it's not very nice @wrkettlitz , as it doesn't make it immediately clear what the purpose is

**Comment by ViIvanov** [comment]
You can try something like this:
```cs
  public abstract class CustomAllocator<T, TMemoryOwner> : MemoryPool<T> where TMemoryOwner : IMemoryOwner<T>
  {
    protected abstract TMemoryOwner RentCore(int minBufferSize = -1);

    public sealed override IMemoryOwner<T> Rent(int minBufferSize = -1) => RentCore(minBufferSize);
  }

  public class FrameAllocator<T> : CustomAllocator<T, FrameAllocator<T>.Frame>
  {
    public struct Frame : IMemoryOwner<T>
    {
      // ...
    }

    protected override Frame RentCore(int minBufferSize = -1) => Rent(minBufferSize);

    public new Frame Rent(int minBufferSize = -1) => new Frame();
  }
```
